Workshop overhaul - Attester section (#285)
KILTprotocol/docs: docs/develop/03_workshop/04_attester/02_did.md

import GenerateKeypairs from '!!raw-loader!@site/code_examples/sdk_examples/src/workshop/attester/generateKeypairs.ts';
import GenerateDid from '!!raw-loader!@site/code_examples/sdk_examples/src/workshop/attester/generateDid.ts';

The next step is to generate a KILT decentralized identifier (DID) using the account you created for the <span className="label-role attester">Attester</span> in [the previous step](./01_account.md).

> <!-- TODO: Also factor into quickstart -->

Reviewer: What do you want to add here?

Author: @Dudleyneedham An internal note to self to check that some of this good intro content was actually in the quickstart

Author: Will move elsewhere

Author: ```suggestion
```

A DID may represent any entity, such as a person, an organization, or a machine.

A DID is a string uniquely identifying each KILT user.
You can store information about a DID on the KILT chain, which is useful for different use cases.

One use case is messaging.

> You could store a public encryption key and a service on-chain, and a user can query both using a DID.

Reviewer 2: I think no `-` is needed in this case, right?

```suggestion
You could store a public encryption key and a service on chain, and a user can query both using a DID.
```

> You could store a public encryption key and a service on-chain, and a user can query both using a DID.

Reviewer 2: I think this would make it a bit more clear.

```suggestion
You could store a public encryption key and a messaging service endpoint, and a user can query both using a DID.
```

## Light and full DIDs

Kilt supports two DID types: **light** and **full**.

There are many differences between the two types, but the most crucial is that you can use a light DID offline, but a full DID needs access to the blockchain to work.
Read the [DID documentation](../../../develop/01_sdk/02_cookbook/01_dids/01_light_did_creation.md) to learn more about the difference between the light and full types.

:::info KILT DID

There are four different key types that a DID supports:

- An _authentication keypair_, used to sign claims and present authenticated credentials 
- A _key-agreement keypair_, used to encrypt/decrypt messages
- An _assertion-method keypair_, used to write CTypes and attestations on chain
- A _capability-delegation keypair_, used to write delegations on chain

You can replace keys over time, e.g., if a key becomes compromised.

:::

## What's the difference between a DID and an account?

A DID and an account sound quite similar, but there are some differences:

- You record both to chain
- You can have a DID without an account
- You can have an account without a DID

> - Only an account can pay deposits and fees and start a DID verification process

Reviewer 2: What is a DID verification process?

- DIDs don't hold any coins

In summary, you register a DID on the blockchain by an account submitting the DID creation transaction and paying the fees.

## Create a DID

As an <span className="label-role attester">Attester</span> needs to interact with the chain, you must create a full DID.

### Generate keypairs

> An <span className="label-role attester">Attester</span> needs all four types of key.

Reviewer 2: This is not technically true. An attester only needs an authentication and an attestation key, at the very minimum.

Reviewer: I guess we only need the first three keys. But maybe we should show generation of all keys.

> Since three of the key types write and sign to the chain, you can use the same key for them using the default KILT keyring to generate them, which is the same keyring used to generate accounts.

Reviewer 2: What does `write and sign to the chain` to the chain mean?

Add the following code to the `attester/generateKeypairs` file.

<TsJsBlock fileName="attester/generateKeypairs">
  {GenerateKeypairs}
</TsJsBlock>

> <!-- TODO: Is this enough? -->

Author: Does the following description describe enough detail?

Reviewer 2: I think it's too much details, for a workshop, if you ask me.

Author: @ntn-x2 Ahh, but you know the subject ;) A workshop is typically pretty detailed…

> Throughout the code are `account.derive` methods that use key derivation syntax. You can read more about this syntax in [the substrate documentation](https://docs.substrate.io/reference/command-line-tools/subkey/#working-with-derived-keys).

Reviewer 2: ```suggestion
Throughout the code are `account.derive` methods that use key derivation syntax. You can read more about this syntax in [the Substrate documentation](https://docs.substrate.io/reference/command-line-tools/subkey/#working-with-derived-keys).
```

The `generateKeypairs` function code derives base and sub keys from a particular path relevant to the use case for each key.
It uses the sr25519 key type, which is the default key type for KILT.

This method works for three of the four key types needed, so the `generateKeyAgreement` function helps generate the key-agreement keypair using the mnemonic.
The function takes the mnemonic and creates another key pair from it using the `sr25519PairFromSeed(mnemonicToMiniSecret(mnemonic))` combination of functions.
The function then creates a secret key based on the earlier temporary key pair and a derivation path relevant to key agreement.

The function returns the key pair needed by generating one more key pair suitable for encryption and decryption using the secret key.

### Write DID to chain

Once you have created all the necessary keys for a DID, you can create the on-chain DID.
To create a DID, load the account created in the [last section](./01_account.md) and use it to pay for the DID registration.
Create and submit the extrinsic (aka transaction) that registers the DID.

<TsJsBlock fileName="attester/generateDid">
  {GenerateDid}
</TsJsBlock>

> The `createFullDid` function takes the key pair generated for the submitter in the previous step and creates a full DID. It returns a mnemonic as a string and DID definition document.

Reviewer 2: Is it key pair or keypair?

> The `createFullDid` function takes the key pair generated for the submitter in the previous step and creates a full DID. It returns a mnemonic as a string and DID definition document.

Reviewer 2: `DID definition document` -> DID Document

Author: ```suggestion
The `createFullDid` function takes the key pair generated for the submitter in the previous step and creates a full DID. It returns a mnemonic as a string and DID document.
```

> Inside the function, the `getStoreTx` method creates A DID creation operation based on the four key pairs created earlier.

Reviewer 2: ```suggestion
Inside the function, the `getStoreTx` method creates a DID creation operation based on the four key pairs created earlier.
```
